storage/rocksdb/ha_rocksdb.cc

```diff
@@ -9338,140 +9253,297 @@ int ha_rocksdb::index_read_map(uchar *const buf, const uchar *const key,
 
     if (!should_recreate_snapshot(rc, is_new_snapshot)) {
       break; /* Exit the loop */
     }
 
     // release the snapshot and iterator so they will be regenerated
     tx->release_snapshot();
-    release_scan_iterator();
+    m_iterator->reset();
   }
 
-  if (rc) {
-    /*
-      This status is returned on any error
-      the only possible error condition is record-not-found
-    */
-    table->m_status = STATUS_NOT_FOUND;
-  } else {
-    table->m_status = 0;
+  if (!rc) {
     /* TODO(yzha) - row stats are gone in 8.0
-    stats.rows_read++;
-    stats.rows_index_first++; */
-    update_row_stats(ROWS_READ);
+    stats.rows_index_first++;
+    stats.rows_index_next--; */
   }
 
   DBUG_RETURN(rc);
 }
 
+
 /*
   @brief
-  Scan the secondary index until we find an index record that satisfies ICP
-
-  @param move_forward   true  <=> move m_scan_it forward
-                        false <=> move m_scan_it backward
-  @param buf            Record buffer (must be the same buffer that
-                        pushed index condition points to, in practice
-                        it is table->record[0])
+    Compute the range lock endpoints and set the range lock, if necessary
+   
+  @param  use_locking_iter OUT  If true, locks are not set and LockingIterator
+                                should be used instead
 
   @detail
-  Move the current iterator m_scan_it until we get an index tuple that
-  satisfies the pushed Index Condition.
-  (if there is no pushed index condition, return right away)
+    If the scanned range doesn't have the endpoint we're scanning towards,
+    don't set the lock, it will be too coarse. Indicate that LockingIterator
+    should be used, instead.
 
   @return
-    0     - Index tuple satisfies ICP, can do index read.
-    other - error code
+     0      Ok
+     Other  Error acquiring the lock (wait timeout, deadlock, etc)
 */
 
-int ha_rocksdb::find_icp_matching_index_rec(const bool move_forward,
-                                            uchar *const buf) {
-  if (pushed_idx_cond && pushed_idx_cond_keyno == active_index) {
-    const Rdb_key_def &kd = *m_key_descr_arr[active_index];
-    THD *thd = ha_thd();
+int ha_rocksdb::set_range_lock(Rdb_transaction *tx,
+                               const Rdb_key_def &kd,
+                               const enum ha_rkey_function &find_flag,
+                               const rocksdb::Slice &slice_arg,
+                               const key_range *const end_key,
+                               bool *use_locking_iterator)
+{
+  rocksdb::Slice end_slice;
+  uchar end_slice_buf[MAX_KEY_LENGTH];
+  bool start_has_inf_suffix = false, end_has_inf_suffix = false;
+  rocksdb::Slice slice(slice_arg);
+  *use_locking_iterator= false;
 
-    while (1) {
-      int rc = rocksdb_skip_expired_records(kd, m_scan_it, !move_forward);
-      if (rc != HA_EXIT_SUCCESS) {
-        return rc;
-      }
+  if (m_lock_rows == RDB_LOCK_NONE || !rocksdb_use_range_locking) {
```

Comment:
For read-committed isolation level, we should probably skip range locks? It seems InnoDB read-committed isolation level doesn't hold range locks (except in short instances like inserting a new row).

Reply:
Yes, done.